### PR TITLE
Make inside ticklabels autorange work with scaleanchor or matches

### PR DESCRIPTION
**Pull request by archmoj**
Fix the bug that causes infinite axes redraw.

@plotly/plotly_js

### CODE REVIEW

**Comment by alexcjohnson** [comment]
Good find!

I don't think we're getting the constraints quite right though - and unfortunately I'm worried that we may need to bake the label adjustment in at a deeper level, as constrained autorange now happens inside `concatExtremes`:

https://github.com/plotly/plotly.js/blob/0dd54673a89e26fdb758e5c11c5cefc4cc25fca8/src/plots/cartesian/autorange.js#L349

That's probably only part of the solution though, as that's just for `matches` constraints. `scaleanchor` constraints are handled in `constraints.enforce`:

https://github.com/plotly/plotly.js/blob/0dd54673a89e26fdb758e5c11c5cefc4cc25fca8/src/plots/cartesian/constraints.js#L547-L549

`axes_chain_scaleanchor_matches2_inside-ticklabels.png` perhaps shows the issue most clearly: one of the features of `axes_chain_scaleanchor_matches2.png` that's really useful is comparing successive rows, green vs red, and black vs blue. The left-most red & green subplots should be identical, as should the right-most blue and black, and in both cases the domain-constrained subplots should still fill the available space (dotted rectangle), because within the subplot there's only a `matches` constraint, which cannot reduce a domain. Yet both the left-most green and right-most black in the inside version have constrained domains, and in the green one this even clips one of the data points.

From one subplot to the next there's a `scaleanchor` constraint, so subsequent green subplots (and earlier black subplots) should end up with smaller domains, which we do see in the new one. But in `axes_chain_scaleanchor_matches2.png` the data points in each green subplot are positioned exactly where the red data points are, and same blue vs black, which I believe should still be the case for the inside case, which mostly looks borne out in the inside case, but not for the first green subplot.

`ticklabelposition3.png` and `axes_chain_scaleanchor_matches_inside-ticklabels.png` have their x labels overlapping the image / heatmaps a bit, which I suspect is a manifestation of the same issue.

File: src/plots/cartesian/axes.js [hunk 1]
```diff
@@ -63,6 +63,7 @@ axes.list = axisIds.list;
 axes.listIds = axisIds.listIds;
 axes.getFromId = axisIds.getFromId;
 axes.getFromTrace = axisIds.getFromTrace;
+axes.isLinked = axisIds.isLinked;
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:66]: Do we need to reexport this here? I think these `Axes` functions are relics of tight coupling to chart studio so should generally not be expanded.

File: src/plots/cartesian/axis_ids.js [hunk 2]
```diff
@@ -136,3 +136,26 @@ exports.ref2id = function(ar) {
...
 };
+
+
+exports.isLinked = function(fullLayout, axId) {
+    var linked = false;
+
+    (fullLayout._axisConstraintGroups || []).forEach(function(e) {
+        if(e[axId]) {
+            linked = true;
+            return;
```
Comment (alexcjohnson) [inline src/plots/cartesian/axis_ids.js:147]: This `return` (and in the next loop) don't do anything... maybe make the `forEach` calls into `for` loops`?

**Review by alexcjohnson — APPROVED** [review]
Looks great! Just 2 small comments and this is ready to go 💃

File: src/plots/cartesian/axis_ids.js (hunk 2 above)
Comment (archmoj) [reply src/plots/cartesian/axis_ids.js]: Good call. Revised in https://github.com/plotly/plotly.js/pull/5329/commits/2bc8420367b84d9b788ac95582650f807f1187cb.

**Pull request merged**